internal/plugin/resolver.go

```diff
@@ -81,30 +81,70 @@ func (resolver *Resolver) Resolve(credentials []*config_v2.Credential) (map[stri
 	errorStrings := make([]string, 0, len(credentials))
 
 	var err error
+
+	// Group credentials by provider
+	var credentialsByProvider = make(map[string][]*config_v2.Credential)
 	for _, credential := range credentials {
-		var provider plugin_v1.Provider
-		var value []byte
+		credentialsByProvider[credential.From] = append(
+			credentialsByProvider[credential.From],
+			credential,
+		)
+	}
 
-		if provider, err = resolver.Provider(credential.From); err != nil {
-			resolver.LogFatalf("ERROR: Provider '%s' could not be used! %v", credential.From, err)
+	// Resolve credentials by provider
+	for providerID, credentialsForProvider := range credentialsByProvider {
+		provider, err := resolver.Provider(providerID)
+		if err != nil {
+			resolver.LogFatalf("ERROR: Provider '%s' could not be used! %v", providerID, err)
 		}
 
-		// This provider cannot resolve the named credential
-		if value, err = provider.GetValue(credential.Get); err != nil {
-			errInfo := fmt.Sprintf("ERROR: Resolving credential '%s' from provider '%s' failed: %v",
-				credential.Get,
-				credential.From,
-				err)
+		// Create secretIds slice and credentialBySecretId map
+		var secretIds = make([]string, len(credentialsForProvider))
+		for idx, cred := range credentialsForProvider {
+			secretIds[idx] = cred.Get
+		}
+
+		// Resolves all credentials for current provider
+		providerResponses, err := provider.GetValues(secretIds...)
+		if err != nil {
+			errInfo := fmt.Sprintf(
+				"ERROR: Resolving credentials from provider '%s' failed: %v",
+				provider.GetName(),
+				err,
+			)
 			log.Println(errInfo)
 
 			errorStrings = append(errorStrings, errInfo)
 			continue
 		}
 
-		result[credential.Name] = value
+		// Collect errors from provider responses
+		var hasErrors bool
+		for _, providerResponse := range providerResponses {
+			if providerResponse.Error != nil {
+				hasErrors = true
+				errorStrings = append(errorStrings, providerResponse.Error.Error())
+				continue
+			}
+		}
+		if hasErrors {
```

Comment:
This brings an interesting question: do we care about finishing resolution of all providers if we hit an error in any provider? My guess is that a single error in fetching variables is enough to throw an error but I'd like to hear your thoughts on this.

Reply:
I think the developer UX of getting all the useful information about errors would be better here. There's no added cost because secret fetching is delegated to the provider and it is at the provider's discretion to fail fast or not. If we have the additional error info I think we ought to show it.